### PR TITLE
Require lifetime bounds for opaque types in order to allow hidden types to capture said lifetimes

### LINKED ISSUE #96996: type-alias-impl-trait implicitly captures lifetime params
```rust
#![feature(type_alias_impl_trait)]
trait Service {
    type Output;
    fn call(self) -> Self::Output;
}
impl<'a> Service for &'a str {
    type Output = impl Sized;
    fn call(self) -> Self::Output { self }
}
```
This compiles now but it shouldn't. Explicit annotation for captured lifetimes in the opaque type `impl Sized + 'a` is required by the RFC.

I guess this is the root cause of #91601.

@rustbot label F-type_alias_impl_trait T-compiler

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #96996

cc @aliemjay

### CODE REVIEW

File: compiler/rustc_borrowck/src/region_infer/opaque_types.rs [hunk 1]
```diff
@@ -239,8 +242,80 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         // shifting.
         let id_substs = InternalSubsts::identity_for_item(self.tcx, def_id.to_def_id());
         debug!(?id_substs);
-        let map: FxHashMap<GenericArg<'tcx>, GenericArg<'tcx>> =
-            substs.iter().enumerate().map(|(index, subst)| (subst, id_substs[index])).collect();
+
+        let map = substs.iter().enumerate().map(|(index, subst)| (subst, id_substs[index]));
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/region_infer/opaque_types.rs:246]: is this just `iter::zip`?

Comment (oli-obk) [reply compiler/rustc_borrowck/src/region_infer/opaque_types.rs:246]: Lol yea, I didn't question the existing code.

**Comment by jackh726** [comment]
@bors r+ rollup=never

While unstable, this a "breaking change" for TAITs, so I'd rather this not land in a rollup, so bisection will be easier for the inevitable issues filed.

**Pull request merged**